### PR TITLE
feat(output): Generate dot file and support multi outformat.(#441)

### PR DESCRIPTION
**Pull request by gabriel-tessier**
Add "dot" in outformat. 
Add support to list for outformat parameter.
Add unit test and documentation.

### CODE REVIEW

**Comment by mingrammer** [comment]
There is already a PR that tries to preserve the `dot` file: https://github.com/mingrammer/diagrams/pull/182.

As I said in that PR, `outformat=dot` may make users confused a bit due to it has different purposes other than the other image format options. preserve_dot_file option would be enough.

I think `preserve_dot_file = True | False` option is better in this case. WDYT?

**Comment by gabriel-tessier** [comment]
I checked the PR 182 and I get your thought! 

In my opinion a user build a Diagram class and by default it generate a png file.
Basically he run python my_diagram.py and he get a xyz.png by specifying the outformat to jpg he get a "jgp", "pdf" or "svg" file. If he want a dot file why he need to use a different option? for me it's not natural.

with Diagram("Simple Diagram", outformat="dot"): ---> if this command generate a dot format is more natural.
than with Diagram("Simple Diagram", preserve_dot_file=True)

Also if you check the output format help page of graphviz https://graphviz.org/doc/info/output.html
there's png, pdf, dot, gv  etc...
So for me it's natural to have a dot file as an output format.

**Comment by gabriel-tessier** [comment]
@mingrammer 

I run the following example and generate the dot file calling render with output "dot" and I also kept the file generated when calling preserve_dot_file.

Here the example:

```python
from diagrams import Diagram
from diagrams.aws.compute import EC2
from diagrams.aws.database import RDS
from diagrams.aws.network import ELB

with Diagram("Grouped Workers", show=False, direction="TB", outformat="dot"):
    ELB("lb") >> [EC2("worker1"),
                  EC2("worker2"),
                  EC2("worker3"),
                  EC2("worker4"),
                  EC2("worker5")] >> RDS("events")

```
Here the output, notice the file name generated with extension: "grouped_workers.dot"

```
digraph "Grouped Workers" {
	graph [bb="0,0,677,544",
		fontcolor="#2D3436",
		fontname="Sans-Serif",
		fontsize=15,
		label="Grouped Workers",
		lheight=0.24,
		lp="338.5,12.5",
		lwidth=1.72,
		nodesep=0.60,
		pad=2.0,
		rankdir=TB,
		ranksep=0.75,
		splines=ortho
	];
	node [fixedsize=true,
		fontcolor="#2D3436",
		fontname="Sans-Serif",
		fontsize=13,
		height=1.4,
		imagescale=true,
		label="\N",
		labelloc=b,
		shape=box,
		style=rounded,
		width=1.4
	];
	edge [color="#7B8894"];
	"9170a8b281b54f11bea106afaab91a6c"	[height=1.9028,
		image="/usr/src/diagrams/resources/aws/network/elastic-load-balancing.png",
		label=lb,
		pos="338.5,475.5",
		shape=none,
		width=1.4028];
	"746b13423bb8492b9fd33ee194801684"	[height=1.9028,
		image="/usr/src/diagrams/resources/aws/compute/ec2.png",
		label=worker1,
		pos="50.5,284.5",
		shape=none,
		width=1.4028];
	"9170a8b281b54f11bea106afaab91a6c" -> "746b13423bb8492b9fd33ee194801684"	[dir=forward,
		fontcolor="#2D3436",
		fontname="Sans-Serif",
		fontsize=13,
		pos="e,50.5,353.23 287.67,498 205,498 50.5,498 50.5,498 50.5,498 50.5,363.23 50.5,363.23"];
	"29bee7ae06804d8f8cda0940aaac7970"	[height=1.9028,
		image="/usr/src/diagrams/resources/aws/compute/ec2.png",
		label=worker2,
		pos="194.5,284.5",
		shape=none,
		width=1.4028];
	"9170a8b281b54f11bea106afaab91a6c" -> "29bee7ae06804d8f8cda0940aaac7970"	[dir=forward,
		fontcolor="#2D3436",
		fontname="Sans-Serif",
		fontsize=13,
		pos="e,194.5,353.11 287.73,452 246.27,452 194.5,452 194.5,452 194.5,452 194.5,363.11 194.5,363.11"];
	"387d0c5fd04b47938f15c0988760f3ec"	[height=1.9028,
		image="/usr/src/diagrams/resources/aws/compute/ec2.png",
		label=worker3,
		pos="338.5,284.5",
		shape=none,
		width=1.4028];
	"9170a8b281b54f11bea106afaab91a6c" -> "387d0c5fd04b47938f15c0988760f3ec"	[dir=forward,
		fontcolor="#2D3436",
		fontname="Sans-Serif",
		fontsize=13,
		pos="e,338.5,353.22 338.5,406.81 338.5,406.81 338.5,363.22 338.5,363.22"];
	f6dedc27548c498c868df2ee5aeece1c	[height=1.9028,
		image="/usr/src/diagrams/resources/aws/compute/ec2.png",
		label=worker4,
		pos="482.5,284.5",
		shape=none,
		width=1.4028];
	"9170a8b281b54f11bea106afaab91a6c" -> f6dedc27548c498c868df2ee5aeece1c	[dir=forward,
		fontcolor="#2D3436",
		fontname="Sans-Serif",
		fontsize=13,
		pos="e,482.5,353.11 389.27,452 430.73,452 482.5,452 482.5,452 482.5,452 482.5,363.11 482.5,363.11"];
	af1c988fa7f04fd3a562fed87a83d1a6	[height=1.9028,
		image="/usr/src/diagrams/resources/aws/compute/ec2.png",
		label=worker5,
		pos="626.5,284.5",
		shape=none,
		width=1.4028];
	"9170a8b281b54f11bea106afaab91a6c" -> af1c988fa7f04fd3a562fed87a83d1a6	[dir=forward,
		fontcolor="#2D3436",
		fontname="Sans-Serif",
		fontsize=13,
		pos="e,626.5,353.23 389.33,498 472,498 626.5,498 626.5,498 626.5,498 626.5,363.23 626.5,363.23"];
	"6a32b36da84b456fa5a1361bc180f3fa"	[height=1.9028,
		image="/usr/src/diagrams/resources/aws/database/rds.png",
		label=events,
		pos="338.5,93.5",
		shape=none,
		width=1.4028];
	"746b13423bb8492b9fd33ee194801684" -> "6a32b36da84b456fa5a1361bc180f3fa"	[dir=forward,
		fontcolor="#2D3436",
		fontname="Sans-Serif",
		fontsize=13,
		pos="e,287.67,70 50.5,215.75 50.5,153.01 50.5,70 50.5,70 50.5,70 277.67,70 277.67,70"];
	"29bee7ae06804d8f8cda0940aaac7970" -> "6a32b36da84b456fa5a1361bc180f3fa"	[dir=forward,
		fontcolor="#2D3436",
		fontname="Sans-Serif",
		fontsize=13,
		pos="e,287.73,116 194.5,215.96 194.5,169.31 194.5,116 194.5,116 194.5,116 277.73,116 277.73,116"];
	"387d0c5fd04b47938f15c0988760f3ec" -> "6a32b36da84b456fa5a1361bc180f3fa"	[dir=forward,
		fontcolor="#2D3436",
		fontname="Sans-Serif",
		fontsize=13,
		pos="e,338.5,162.22 338.5,215.81 338.5,215.81 338.5,172.22 338.5,172.22"];
	f6dedc27548c498c868df2ee5aeece1c -> "6a32b36da84b456fa5a1361bc180f3fa"	[dir=forward,
		fontcolor="#2D3436",
		fontname="Sans-Serif",
		fontsize=13,
		pos="e,389.27,116 482.5,215.96 482.5,169.31 482.5,116 482.5,116 482.5,116 399.27,116 399.27,116"];
	af1c988fa7f04fd3a562fed87a83d1a6 -> "6a32b36da84b456fa5a1361bc180f3fa"	[dir=forward,
		fontcolor="#2D3436",
		fontname="Sans-Serif",
		fontsize=13,
		pos="e,389.33,70 626.5,215.75 626.5,153.01 626.5,70 626.5,70 626.5,70 399.33,70 399.33,70"];
}

```

Here the file built to generate the file basically the one if we set the option "preserve_dot_file = True", the file generated don't have extension: "grouped_workers" and content is not formatted.

```
digraph "Grouped Workers" {
	graph [fontcolor="#2D3436" fontname="Sans-Serif" fontsize=15 label="Grouped Workers" nodesep=0.60 pad=2.0 rankdir=TB ranksep=0.75 splines=ortho]
	node [fixedsize=true fontcolor="#2D3436" fontname="Sans-Serif" fontsize=13 height=1.4 imagescale=true labelloc=b shape=box style=rounded width=1.4]
	edge [color="#7B8894"]
	"9170a8b281b54f11bea106afaab91a6c" [label=lb height=1.9 image="/usr/src/diagrams/resources/aws/network/elastic-load-balancing.png" shape=none]
	"746b13423bb8492b9fd33ee194801684" [label=worker1 height=1.9 image="/usr/src/diagrams/resources/aws/compute/ec2.png" shape=none]
	"29bee7ae06804d8f8cda0940aaac7970" [label=worker2 height=1.9 image="/usr/src/diagrams/resources/aws/compute/ec2.png" shape=none]
	"387d0c5fd04b47938f15c0988760f3ec" [label=worker3 height=1.9 image="/usr/src/diagrams/resources/aws/compute/ec2.png" shape=none]
	f6dedc27548c498c868df2ee5aeece1c [label=worker4 height=1.9 image="/usr/src/diagrams/resources/aws/compute/ec2.png" shape=none]
	af1c988fa7f04fd3a562fed87a83d1a6 [label=worker5 height=1.9 image="/usr/src/diagrams/resources/aws/compute/ec2.png" shape=none]
	"9170a8b281b54f11bea106afaab91a6c" -> "746b13423bb8492b9fd33ee194801684" [dir=forward fontcolor="#2D3436" fontname="Sans-Serif" fontsize=13]
	"9170a8b281b54f11bea106afaab91a6c" -> "29bee7ae06804d8f8cda0940aaac7970" [dir=forward fontcolor="#2D3436" fontname="Sans-Serif" fontsize=13]
	"9170a8b281b54f11bea106afaab91a6c" -> "387d0c5fd04b47938f15c0988760f3ec" [dir=forward fontcolor="#2D3436" fontname="Sans-Serif" fontsize=13]
	"9170a8b281b54f11bea106afaab91a6c" -> f6dedc27548c498c868df2ee5aeece1c [dir=forward fontcolor="#2D3436" fontname="Sans-Serif" fontsize=13]
	"9170a8b281b54f11bea106afaab91a6c" -> af1c988fa7f04fd3a562fed87a83d1a6 [dir=forward fontcolor="#2D3436" fontname="Sans-Serif" fontsize=13]
	"6a32b36da84b456fa5a1361bc180f3fa" [label=events height=1.9 image="/usr/src/diagrams/resources/aws/database/rds.png" shape=none]
	"746b13423bb8492b9fd33ee194801684" -> "6a32b36da84b456fa5a1361bc180f3fa" [dir=forward fontcolor="#2D3436" fontname="Sans-Serif" fontsize=13]
	"29bee7ae06804d8f8cda0940aaac7970" -> "6a32b36da84b456fa5a1361bc180f3fa" [dir=forward fontcolor="#2D3436" fontname="Sans-Serif" fontsize=13]
	"387d0c5fd04b47938f15c0988760f3ec" -> "6a32b36da84b456fa5a1361bc180f3fa" [dir=forward fontcolor="#2D3436" fontname="Sans-Serif" fontsize=13]
	f6dedc27548c498c868df2ee5aeece1c -> "6a32b36da84b456fa5a1361bc180f3fa" [dir=forward fontcolor="#2D3436" fontname="Sans-Serif" fontsize=13]
	af1c988fa7f04fd3a562fed87a83d1a6 -> "6a32b36da84b456fa5a1361bc180f3fa" [dir=forward fontcolor="#2D3436" fontname="Sans-Serif" fontsize=13]
}

```

In any case if you prefer to set a new parameter 'preserve_dot_file' or if you consider the "dot" in the outformat (Important is that it's correctly documented).
I think it's better to generate the dot file with the extension and formatted by calling the render() method, so the original file can be removed.

WDYT? 

If the way defined in the PR 187 is better for you I can close this PR and make a new one with only multi outformat.
I made 2 changes in one PR as I follow the issue which had 2 suggestion changes.

File: diagrams/__init__.py [hunk 1]
```diff
@@ -188,7 +190,11 @@ def subgraph(self, dot: Digraph) -> None:
         self.dot.subgraph(dot)
 
     def render(self) -> None:
-        self.dot.render(format=self.outformat, view=self.show, quiet=True)
+        if isinstance(self.outformat, list):
+          for one_format in self.outformat:
+            self.dot.render(format=one_format, view=self.show, quiet=True)
+        else:
+          self.dot.render(format=self.outformat, view=self.show, quiet=True)
```
Comment (mingrammer) [inline diagrams/__init__.py:197]: 4 indents please :)

**Comment by mingrammer** [comment]
> Also if you check the output format help page of graphviz https://graphviz.org/doc/info/output.html
> there's png, pdf, dot, gv etc...
> So for me it's natural to have a dot file as an output format.

It makes sense. `dot` outformat looks more natural to me too.

**Comment by tonyyzy** [comment]
Hey all,  what's the status on getting this merged? (other than the indentation

**Comment by gabriel-tessier** [comment]
@tonyyzy 

Ohhh!! I mess the indent part!! 
I will push a fix.

**Comment by gabriel-tessier** [comment]
@mingrammer 
I had problem with my branch to apply the changes so I create a new PR:

https://github.com/mingrammer/diagrams/pull/592

I close this one as the other one have the same commits + the indentation fix.

**Pull request closed without merging**